DM-38124: Improve configuration to support pluggable controllers

couger01 opened the pull request.

couger01 committed:
Improve configuration to support pluggable controllers

r-owen commented on python/lsst/ts/linearstage/config_schema.py:
```diff
@@ -2,103 +2,62 @@
         type: number
-    steps_per_mm:
-        description: This is approximately the amount of steps in a millimeter (for the Zaber stage only).
+      steps_per_mm:
```
Every instance of
```
type: object
properties:
```
Should *also* have `required` and `addtionalProperties: false` at the same level. We usually put this after the last property is defined (though I don't think that is necessary). This catches two very common sources of configuration errors: missing items and misspelled items.

I find 3 places where this is wanted. One of them is here, after the last field (line 39) please add:
```
required:
  - serial_port
  - daisy_chain_address
  - ...
additionalProperties: false
```

r-owen commented on python/lsst/ts/linearstage/controllers/igus_dryve/igus_dryve.py:
```diff
@@ -3,6 +3,7 @@
 import asyncio
 import logging
 import time
+import types
```
At line 18 below please delete basicConfig.

Never mess with global logging configuration in library code. It can make things very confusing for anyone trying to understand your code.

Please also look for this in other code in this package. At the executable level it's fine if you really want it (and it's fairly common in unit tests). I'm only saying to eliminate it from code that is imported (library code).

r-owen commented on python/lsst/ts/linearstage/controllers/igus_dryve/igus_dryve.py:
```diff
@@ -83,17 +84,18 @@ def __init__(self, simulation_mode, log) -> None:
     def configure(self, config):
-        self.port = config.socket_port
-        self.address = config.socket_address
+        config.igus = types.SimpleNamespace(**config.igus)
```
I very strongly strongly suggest not modifying the object passed in. It makes the code harder to understand (most users will assume the config is read only) and it is not needed in this case.

Since this class *only* needs config.igus entries, I recommend the following:
```
self.config = types.SimpleNamespace(**config.igus)
```
It has these advantages:
* It avoids copying each field by name.
* It simplifies future changes to the schema, since you don't have to remember to update these lines.
* It groups related data together and reduces the number of minor instance variables.

If you also need one or two other config items, it gets a bit messier, but there are ways to deal with it.

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst/zaber_lst.py:
```diff
@@ -134,10 +135,11 @@ def configure(self, config):
         config : `types.SimpleNamespace`
         """
-        self.steps_conversion = config.steps_per_mm
+        config.zaber = types.SimpleNamespace(**config.zaber)
```
Same comment as for igus config handling above. In this case you have fewer config variables, but I still suggest saving the config as instance variable `config`.

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst/zaber_lst.py:
```diff
@@ -134,10 +135,11 @@ def configure(self, config):
-            self.port = config.serial_port
-            self.address = config.daisy_chain_address
+            self.port = config.zaber.serial_port
+            self.address = config.zaber.daisy_chain_address
```
I suggest eliminating the conditional and always setting port and address, or, as suggested above, setting self.config = ... , which means self.config will include the configured port and address.

It's true that in simulation mode you ignore these values, but the user can still see what the config specified. And it simplifies the code.

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst/zaber_lst.py:
```diff
@@ -4,6 +4,7 @@
 import logging
 import os
 import pty
+import types
```
This comment is about the whole file, not this particular line of it:

Please move this file up one level and eliminate the zabar_lst directory. It simplifies the layout and you only have one file which defines the Zabar code, so it's not as if you have to keep a bunch of related sub-modules together.

r-owen replied:
Please do this.

r-owen commented on python/lsst/ts/linearstage/csc.py:
```diff
@@ -39,6 +40,8 @@ def __init__(
         config_dir=None,
         simulation_mode=0,
+        override=None,
+        **kwargs,
```
What are these extra kwargs? I don't think there are any. Please consider deleting **kwargs.

r-owen replied:
Ping.

r-owen commented on python/lsst/ts/linearstage/csc.py:
```diff
@@ -69,24 +74,32 @@ async def configure(self, config):
-            )
+        for index in range(self.salinfo.index):
+            if self.salinfo.index == config.linear_stage_config[index]["sal_index"]:
+                config = types.SimpleNamespace(**config.linear_stage_config[index])
```
This shadows the `config` argument, which is confusing for readers and potentially buggy. Please use a different variable name, or just skip the conversion and use the directly via dict notation.

r-owen commented on python/lsst/ts/linearstage/csc.py:
```diff
@@ -69,24 +74,32 @@ async def configure(self, config):
-            self.component = ZaberLSTStage(
-                simulation_mode=bool(self.simulation_mode_number), log=self.log
-            )
+        for index in range(self.salinfo.index):
```
I think you need to iterate over config.linear_stage_config (or the number of entries in it) to reliably find the config. For example:
```
for linear_stage_config in config.linear_stage_config:
    if self.salinfo.index !=  linear_stage_config["sal_index"]:
        continue
    self.stage_type = linear_stage_config["stage_type"]
    self.target_position_minimum = linear_stage_config["target_position_minimum"]
    self.target_position_maximum = linear_stage_config["target_position_maximum"]
    ....
else:
    raise...
```
where I have also not bothered to create a types.SimpleNamespace because there are so few items to process and because it needs yet another variable name.

r-owen commented on python/lsst/ts/linearstage/csc.py:
```diff
@@ -69,24 +74,32 @@ async def configure(self, config):
+                )
+                # Instantiate the class specific to the hardware component
+                if self.stage_type == "Igus":
+                    self.component = IgusLinearStageStepper(
```
What happens if the CSC is sent to STANDBY and back to DISABLED? This method is called again. Surely you need to somehow shut down the old self.component? I don't see any code to do that. For instance in handle_summary_state or somewhere similar, you could close the old component and perhaps set it to None.

r-owen commented on python/lsst/ts/linearstage/csc.py:
```diff
@@ -69,24 +74,32 @@ async def configure(self, config):
+                else:
+                    raise IOError("Stage type not defined in config file.")
+            else:
+                raise RuntimeError("Configuration not found for SAL index.")
```
I strongly suggest passing the configuration as part of the constructor (to both linear stage classes). This offers several important advantages:
* It is impossible to construct a mis-configured component. If the config is invalid this is caught in the constructor itself.
* It is impossible to try to use a component before it has been configured.

Thus it simplifies the API and makes it easier to use correctly and  harder to use incorrectly.

r-owen commented on tests/data/config/_init.yaml:
```diff
@@ -1,10 +1,23 @@
-target_position_maximum: 70
+linear_stage_config:
+  -
+    sal_index: 1
```
I gently suggest combining the above two lines, resulting in:
```
linear_stage_config:
  - sal_index: 1
    target_position_minimum: 5
    ...
```
(if your code is the result of an auto formatter then I withdraw the suggestion).

r-owen commented on tests/data/config/igus.yaml:
```diff
@@ -2,15 +2,21 @@
+      homing_acceleration: 20
+      homing_timeout: 60
+      motion_speed: 50
+      motion_acceleration: 20
```
Alas this shows the annoyance with overriding pluggable algorithms for indexed SAL components.

If you make igus and zabar required, as I suggest elsewhere, then you will have to include the zabar config here, else this will fail validation.

Sorry, but multiple algorithms coupled with multiple SAL indices is a nuisance.

Two other options are as follows:

1: Make `igus` and `zabar` optional fields (but still add `addtionalProperties: false` to that section of the schema so users can't specify things like "igs" or "zabbar") and then inside the CSC's `configure` method check if the required "igus" or "zabar" field is present and raise an informative salobj.ExpectedError if it is not.

I think this would be pretty clean, but it _may_ violate our rules on configuration (I am not sure), so if you do want to take this route please talk to Tiago about it first. I will be happy to join that conversation if you want. I hope it is allowed, because is rather similar to the next option, which is definitely permitted.

2: Make the two stage classes specify their own configuration schema, rather than putting that information in the CSC's schema. Then specify stage-specific config using a *generic* required field such as `stage_config`.

This is the approach used by most of our CSCs that have pluggable algorithms. It has a few advantages:
* The device-specific config can have default values.
* Overrides are a bit simpler (no need to specify both `igus` and `zabar` here).
* Each sal-index-specific piece of configuration only has config for the device actually being used, making it easier to read.
* It scales well to larger numbers of devices (not important for your case).

But it also has a few disadvantages:
* It can be harder for a reader to understand what the available config fields are; you can't necessarily just read the config files to figure it out (though if the config file has config for both kinds of stages then that is not much of a problem -- especially if you forego default values for your fields).
* It is a bit more work. The stage-specific schema has to be defined inside each stage class (e.g. by adding a get_config_schema classmethod). And the CSC has to retrieve the schema and validate it manually (by constructing a salobj.DefaultingValidator, if you want to allow default values, else you can use the standard jsonschema validator).

See ts_ess_csc (data-client-specific config) or ts_watcher (rule-specific config) for examples.

couger01 replied:
I chose option 2.

r-owen commented on tests/data/config/zaber.yaml:
```diff
@@ -1,10 +1,13 @@
-steps_per_mm: 8000
-target_position_minimum: 5
-target_position_maximum: 70
+linear_stage_config:
```
Same comment as for igus.yaml above

r-owen commented on python/lsst/ts/linearstage/config_schema.py:
```diff
@@ -2,103 +2,62 @@
-        type: number
-    serial_port:
-        description: USB port for the serial interface
+  linear_stage_config:
```
Please name this field `instances`. That matches reality very well: it is the configuration for one instance  (one specific SAL index) of your CSC and is what other similar CSCs do.

r-owen reviewed (changes requested):
This is a lot closer, but still has a few issues, including:
* I think the way you iterate over sal-index-specific configs is not quite right.
* The schema is missing `required` and `additionalProperties` fields in 3 places. And once you add these I think you will find two of your tests schemas no longer pass validation. I give a few options for fixing that, some of which require software changes.

couger01 committed:
Make configuration tied to sal index

* Restructure zaber and igus classes
* Use ABC for stages
* Use sub schemas defined in classes to validate configuration files

r-owen commented on python/lsst/ts/linearstage/config_schema.py:
```diff
@@ -5,100 +5,34 @@
-        - steps_per_mm
-
+        - stage_type
+        - stage
```
Please add `additionalProperties: false` here, at the same indentation level as `required:` on line 28 above.

r-owen commented on python/lsst/ts/linearstage/config_schema.py:
```diff
@@ -5,100 +5,34 @@
+            enum:
+                - Zaber
+                - Igus
+        stage:
```
I gently suggest `stage_config`, to reduce ambiguity.

r-owen commented on python/lsst/ts/linearstage/controllers/icontroller.py:
```diff
@@ -0,0 +1,45 @@
+from abc import ABC, abstractmethod
+
+
+class Stage(ABC):
```
Please add a doc string.

Please consider adding a concrete `__init__` method that sets accepts the config, log, and simulation_mode arguments and sets attributes of the same name.

r-owen commented on python/lsst/ts/linearstage/controllers/igus_dryve/igus_dryve.py:
```diff
@@ -77,23 +76,25 @@ def __init__(self, simulation_mode, log) -> None:
     def configure(self, config):
-        self.port = config.socket_port
-        self.address = config.socket_address
+        self.config = types.SimpleNamespace(**config["stage"])
```
Please document that `config` is the full instance configuration.

Or if all your subclasses of Stage need only the stage-specific config (and you are sure that will not change), consider just passing in the stage-specific config.

r-owen commented on python/lsst/ts/linearstage/controllers/igus_dryve/igus_dryve.py:
```diff
@@ -77,23 +76,25 @@ def __init__(self, simulation_mode, log) -> None:
+        self.homing_timeout = self.config.homing_timeout
+        self.motion_speed = self.config.motion_speed
+        self.motion_acceleration = self.config.motion_acceleration
+        self.maximum_stroke = self.config.maximum_stroke
```
I *strongly* suggest not copying the config. Just save the config and get these properties from self.config when you need them. Less duplication, just one way to do things, and robust against config tweaks.

r-owen commented on python/lsst/ts/linearstage/controllers/igus_dryve/igus_dryve.py:
```diff
@@ -1087,3 +1084,49 @@ async def send_telegram(
+            description: Sub-schema for Igus configuration files
+            type: object
+            properties:
+                igus_controller:
```
This property not needed. We already know this config is for the igus controller. I strongly suggest deleting it and moving the remaining entries up one level.

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst/zaber_lst.py:
```diff
@@ -123,45 +123,48 @@ def __init__(self, simulation_mode, log) -> None:
-            self.port = config.serial_port
-            self.address = config.daisy_chain_address
+            self.port = self.config.serial_port
+            self.address = self.config.daisy_chain_address
```
Please don't duplicate the info. Just use self.config.x when you want config item x.

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst/zaber_lst.py:
```diff
@@ -418,3 +421,26 @@ async def stop(self):
+        description: Schema for LinearStage configuration files
+        type: object
+        properties:
+            zaber_controiller:
```
As for igus, you don't need this property. Please delete it and move the other items up a level.

r-owen reviewed (changes requested):
I have left some comments and found others that have yet to be implemented. For each comment please either implement it and mark it resolved, or respond to it in some other way (such as explaining why you would rather not do that, or what you are going to do instead).

Also in looking at the code directly I see a number of instances of `tuple([...])`. I suggest changing these to `(...)` because it is easier to read, saves an indentation level for tuples that are too long to fit on one line, and is more idiomatic.

r-owen commented on python/lsst/ts/linearstage/controllers/igus_dryve.py:
```diff
@@ -61,9 +60,8 @@ class IgusLinearStageStepper:
-        self.log = log.getChild("IgusLinearStageStepper")
-        self.connected = False
+    def __init__(self, config, simulation_mode, log) -> None:
+        self.log = logging.getLogger(__name__)
```
You are missing the call to the parent class constructor. I thought that was required, and so I am surprised that your current code passed tests without warnings from Python. But in any case it is certainly wanted, and in this case it allows you to delete many lines of code in this class.

Add this:
```
super().__init__(config=config, simulation_mode= simulation_mode, log=log)
```
And remove the following (since the parent class does it for you):
* self.log =
* self.simulation_mode =
* The `configure` method and your call to it

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst.py:
```diff
@@ -62,9 +62,8 @@ class ZaberLSTStage:
-        self.log = log.getChild("ZaberLSTStage")
-        self.connected = False
+    def __init__(self, config, simulation_mode, log) -> None:
+        self.log = logging.getLogger(__name__)
```
As for the Igus stage, you are missing the call to `super().__init__`. Please see the comment for Igus and do the same things here.

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst.py:
```diff
@@ -123,45 +122,48 @@ def __init__(self, simulation_mode, log) -> None:
-            self.port = config.serial_port
-            self.address = config.daisy_chain_address
+            self.port = self.config.serial_port
+            self.address = self.config.daisy_chain_address
```
Please delete this method and:
* Delete self.steps_conversion and use `self.config.steps_per_mm` directly.
* Delete self.port and self.address; get the port and address straight from the config.

Deleting those attributes avoids duplicating information and in the first case it also avoids the confusion of calling one thing by two different names.

r-owen replied:
Thank you for deleting the contents of the method. But please still delete the method. The base class already assigns self.config=config so there is no need to repeat it in the subclasses, and thus no need for a `configure` method.

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst.py:
```diff
@@ -174,10 +176,10 @@ async def enable_motor(self, value):
-        )
+        pass
+
+    async def disable_motor(self):
```
Please add a short doc string -- one that notes that the zabarLST is always enabled.

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst.py:
```diff
@@ -264,7 +266,7 @@ async def move_relative(self, value):
         await asyncio.sleep(_ZABER_MOVEMENT_TIME)
 
-    async def get_home(self):
+    async def home(self):
```
I have a question about line 290: `self.commander.set(cmd)`: how do you know that if it raises SerialException that this is a timeout? Where is the amount of timeout specified?

A code comment would be helpful to explain these things.

Naively it is not at all clear what SerialException means. It sounds like an issue with communication -- e.g. a mismatched baud rate or something.

r-owen replied:
I don't see any change here. In the Raises section of the doc string you describe SerialException as "Raised when the serial port has a problem." which does not sound like a timeout to me.

Is timeout one of the possible problems? If so, what sets the amount of timeout. If not, do you want to add a timeout by wrapping the call in `await asyncio.wait_for(self.commander.set(cmd), timeout=...)`?

In any case, please make the log message and the docs consistent. The doc string says there is a timeout, the log messages says any serial port issue is actually a timeout, but the Raises section of the doc string is much more general (and much closer to what I would expect a SerialException to be).

This same issue applies to both linear stages (and would likely cease to apply if you could use Moxa serial to ethernet converters).

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst.py:
```diff
@@ -418,3 +420,28 @@ async def stop(self):
+        $id: https://github.com/lsst-ts/ts_LinearStage/blob/master/schema/LinearStage.yaml
+        # title must end with one or more spaces followed by the schema version, which must begin with "v"
+        title: Zaber v1
+        description: Schema for LinearStage configuration files
```
This seems to be the original description for the full CSC config?

I suggest adding "Zabar" to the description and unifying the description of this and Igus so they sound more similar. Personally I think of these as stage-specific schemas rather than sub-schemas, so I'd be inclined to use something like "Schema for the Zabar/Igus linear stage".

r-owen commented on python/lsst/ts/linearstage/csc.py:
```diff
@@ -69,26 +71,25 @@ async def configure(self, config):
+        self.log.info(f"{config=}")
+        self.stage_type = instance["stage_type"]
+        module = importlib.import_module("lsst.ts.linearstage.controllers")
+        stage_class = getattr(module, self.stage_type)
```
There is no need to use importlib and it would simplify the code to avoid it. Just import  lsst.ts.linearstage.controllers in the normal way at the top of the file (either as a relative or absolute import) and then call getattr on whatever you import.

In other words, one of these:
```
from . import controllers  # at the top. with other imports
...
stage_class = getattr(controllers, config.stage_type)
```
or
```
import lsst.ts.linearstage.controllers  # at the top, with other imports
...
stage_class = getattr(lsst.ts.linearstage.controllers, config.stage_type)
```

r-owen commented on python/lsst/ts/linearstage/csc.py:
```diff
@@ -69,26 +71,25 @@ async def configure(self, config):
+            else:
+                raise RuntimeError(f"NO config found for {self.salinfo.index=}")
+        self.log.info(f"{config=}")
+        self.stage_type = instance["stage_type"]
```
Please delete self.stage_type and just use `instance["stage_type"]`

or at the very least make it a *local* variable `stage_type = instance["stage_type"]`.

You only use it once, right here, so it seems wasteful to save it as an instance variable.

r-owen commented on tests/test_igus_dryve.py:
```diff
@@ -1,37 +1,44 @@
 
-from lsst.ts.linearstage.controllers.igus_dryve import IgusLinearStageStepper
-from lsst.ts.linearstage.controllers.igus_dryve.telegrams import (  # telegrams_read_errs,
+from lsst.ts.linearstage.controllers.igus_dryve import Igus
```
I suggest omitting `igus_dryve` from this import, since the module name is an internal detail. In fact that applies to all of these imports. This should work fine, since you moved all those modules up out of the igus directory:
```
from lsst.ts.linearstage.controllers import Igus, telegrams_read, telegrams_write, interpret_read_telegram
```

r-owen commented on tests/test_zaber_lst.py:
```diff
@@ -1,17 +1,21 @@
 
 import pytest
-from lsst.ts.linearstage.controllers.zaber_lst import ZaberLSTStage
+from lsst.ts.linearstage.controllers.zaber_lst import Zaber
```
Please omit zabar_lst

r-owen reviewed (changes requested):
Looks much better. The one blocker I found (and I'm sorry I didn't catch it earlier) is failure to call super().__init__ in your concrete stage classes. Fortunately this allows you to delete a method and several lines of code from both classes, so it's a big win.

I also have a few minor suggestions I hope you will consider.

r-owen commented on python/lsst/ts/linearstage/controllers/igus_dryve.py:
```diff
@@ -76,24 +73,19 @@ def __init__(self, simulation_mode, log) -> None:
+        config : `types.Simplenamespace`
+            The full configuration object.
+        """
+        self.config = config
```
Please delete this method. It is a no-op because the base class already sets self.config=config.

r-owen commented on python/lsst/ts/linearstage/controllers/utils.py:
```diff
@@ -1,9 +1,7 @@
-from ...controllers.igus_dryve.telegrams import (  # telegrams_read,; telegrams_read_errs,
-    telegrams_write,
-)
+from .telegrams import telegrams_write  # telegrams_read,; telegrams_read_errs,
```
Please remove the comment. You can always add them later if you need them.

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst.py:
```diff
@@ -121,47 +119,44 @@ def __init__(self, simulation_mode, log) -> None:
-        if not self.simulation_mode:
-            self.port = config.serial_port
-            self.address = config.daisy_chain_address
+        self.config = config
```
Please delete this method. It is a no-op since the base class already does this for you.

r-owen commented on python/lsst/ts/linearstage/controllers/icontroller.py:
```diff
@@ -0,0 +1,71 @@
+from abc import ABC, abstractmethod
```
Please rename this module to `stage.py` so its name reflects its contents.

(This is a comment about the file name, not the line above).

r-owen commented on python/lsst/ts/linearstage/controllers/icontroller.py:
```diff
@@ -0,0 +1,71 @@
+from abc import ABC, abstractmethod
+
```
Please add `__all__ = ["Stage"]` -- before any imports.

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst.py:
```diff
@@ -356,15 +355,16 @@ def get_position(self):
             if status_dictionary:
                 self.log.info("Position captured")
-                return float(float(reply.data) / self.steps_conversion)
+                return float(float(reply.data) / self.config.steps_per_mm)
```
At line 369 below please change:
```
    self.log.error(e)
    self.log.info("Command for device timed out")
```
to something that includes the message and error in the same log message, preferably at the same severity level (presumably `error`), e.g.:

```
    self.log.error("Communication error: {e}")
```
If this is always going to be a timeout, then fine. But I'm skeptical of that and it doesn't match your description in the doc string.

Speaking of timeouts: is SerialException raised in case of a timeout? If so, what parameter controls the length of the timeout. If not, do you want to add an explicit timeout, e.g. with `await asyncio.wait_for(self.commander.send(...), timeout=...)`?

r-owen commented on python/lsst/ts/linearstage/controllers/zaber_lst.py:
```diff
@@ -396,7 +396,7 @@ def retrieve_status(self):
             raise
 
-    async def publish(self):
+    async def update(self):
```
Please do a global search for `self.log.error(e)` and add context.

For example at line 396 above, change this to something like:
```
    self.log.error(f"retrieve_status failed: {e}")
    raise
```
A search turned up 6 instances, all in this module.

r-owen commented on tests/test_csc.py:
```diff
@@ -151,25 +143,37 @@ async def test_moveAbsolute(self):
+                        # behavior should be. So just going to handle this
+                        # specially.
+                        posit = await self.remote.tel_position.next(flush=True)
+                        assert math.isclose(posit.position, 11.2, rel_tol=0.05)
```
I strongly suggest using `assert posit.position == pytest.approx(11.2, abs=0.05)` and similarly instead of `assert math.isclose` because it gives a much clearer assertion message if it fails.

I made a small test file to demonstrate and got these summaries:
FAILED tests/test_temp.py::TestApprox::test_math_isclose - assert False
FAILED tests/test_temp.py::TestApprox::test_pytest_approx - assert 1.1 == 2.2 ± 2.2e-06

r-owen commented on tests/test_csc.py:
```diff
@@ -151,25 +143,37 @@ async def test_moveAbsolute(self):
+                        )
+                    else:
+                        posit = await self.remote.tel_position.next(flush=True)
+                        assert math.isclose(posit.position, 21.2, rel_tol=0.05)
```
Please delete test_checkMotorInternalStatusPreservation below or enable it. If you feel the need to keep it -- e.g. for future work -- then add a Jira ticket to do the work, add a TODO comment here that references the ticket, and (preferably) use unittest.skip with a suitable message, instead of commenting out the code, so it is more obvious that a test is being skipped and why.

couger01 replied:
Not sure what Patrick's intention with it is. However I can ask him about it.

couger01 replied:
Deleted it.

r-owen reviewed (changes requested):
Very close now. Thank you for all the nice improvements.

I still have some requests, including:
* Get rid of the configure methods in the Stage subclasses.
* Change `tuple([...items...])` to `(...items...)` everywhere. If you are using a global search then look for `tuple(` because most of these span multiple lines.
* Be consistent in how you describe SerialException. And will that be raised for a timeout? If so, what sets the timeout time. If not, do you want to add an explicit timeout? This involves reworking some doc strings and log messages to be more self-consistent.

But see the comments for details.

A new thing I found today is that few, if any, .py files have the standard boilerplate header. Please add that (unless we have permission to omit it?).

r-owen committed:
Simplify tuples

Change `tuple([...items...])` to `(...items...)`.
Change `tuple(list(...items...))` to `tuple(...items...)`

r-owen reviewed (commented):
A few more things to fix:

Your use of `TimeoutError` from the timeout error appears to be incorrect. In the Zabar code apparently it should be `zabar.TimeoutError`. In the Igus code I am guessing it should be `asyncio.TimeoutError`.

Some .py files still are missing the boilerplate header. Thank you for adding it to some of them.

Still some `tuple([...])` items to fix. I submitted a PR to clean up the remaining ones I found and assigned you as reviewer. Please merge it if you agree with it.

r-owen committed:
azbar_lst: fix Raises doc referring to the wrong TimeoutError

r-owen committed:
Add missing boilerplate to one module

r-owen committed:
test_csc.py: remove obsolete if __name_ ==...

r-owen reviewed (approved):
Thank you for all the cleanup (including of old issues).

I made a PR of small tweaks that I hope you will consider merging to this branch before merging this branch: https://github.com/lsst-ts/ts_LinearStage/pull/26

Other minor cleanups to consider for now or the future:
* Remove the SConstruct and Sconscript files. We don't use scons anymore.
* Remove ups/ts_linearstage.cfg: it is not needed for a pure Python package
